crates/orchestrator/src/client.rs

```diff
@@ -136,12 +139,52 @@ impl OrchestratorClient {
         };
 
         let mut config = self.wait_for_fn_from_orchestrator(f).await;
-
         config.node_index = From::<u16>::from(node_index);
+        // The orchestrator will generate keys for validator if it doesn't load keys from file
+        config.config.my_own_validator_config =
+            ValidatorConfig::<K>::generated_from_seed_indexed(config.seed, config.node_index, 1);
 
         config
     }
 
+    /// Sends my public key to the orchestrator so that it can collect all public keys
+    /// And get the updated config
+    /// Blocks until the orchestrator collects all peer's public keys/configs
+    /// # Panics
+    /// if unable to post
+    pub async fn post_and_wait_all_public_keys<K: SignatureKey, E: ElectionConfig>(
+        &self,
+        node_index: u64,
+        my_pub_key: K,
+    ) -> NetworkConfig<K, E> {
+        // send my public key
+        let _send_pubkey_ready_f: Result<(), ClientError> = self
+            .client
+            .post(&format!("api/pubkey/{node_index}"))
+            .body_binary(&my_pub_key.to_bytes())
+            .unwrap()
+            .send()
+            .await;
+
+        // wait for all nodes' public keys
+        let wait_for_all_nodes_pub_key = |client: Client<ClientError>| {
+            async move { client.get("api/peer_pub_ready").send().await }.boxed()
+        };
+        self.wait_for_fn_from_orchestrator::<_, _, ()>(wait_for_all_nodes_pub_key)
+            .await;
+
+        // get the newest updated config
+        let get_newest_config = |client: Client<ClientError>| {
+            async move {
+                let config: Result<NetworkConfig<K, E>, ClientError> =
+                    client.get("api/config_after_peer_collected").send().await;
+                config
+            }
+            .boxed()
+        };
+        self.wait_for_fn_from_orchestrator(get_newest_config).await
```

Comment:
Can we just call `send().await` to get the updated config? Is it possible that we need `wait_for_fn_from_orchestrator` though we know peers' keys are ready?

Reply:
Yeah that's redundant.